File: golang/publishing_message_test.go
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package golang

import (
	v2 "github.com/apache/rocketmq-clients/golang/v5/protocol/v2"
	"testing"
)

func TestNewPublishingMessage(t *testing.T) {
	namespace := "ns-test"
	pSetting := &producerSettings{}
	msg := &Message{}
	pMsg, err := NewPublishingMessage(msg, namespace, pSetting, false)
	if err != nil {
		t.Error(err)
	}
	v2Msg, err := pMsg.toProtobuf()
	if err != nil {
		t.Error(err)
	}
	if v2Msg.GetTopic().GetResourceNamespace() != namespace {
		t.Error("namespace not equal")
	}
}

func TestNewPublishingMessage_LiteMessage(t *testing.T) {
	namespace := "ns-test"
	pSetting := &producerSettings{}
	msg := &Message{
		liteTopic: ptrToString("lite-topic"),
	}
	pMsg, err := NewPublishingMessage(msg, namespace, pSetting, false)
	if err != nil {
		t.Error(err)
	}
	if pMsg.messageType != v2.MessageType_LITE {
		t.Errorf("expected messageType to be LITE, got %v", pMsg.messageType)
	}
}

func ptrToString(s string) *string {
	return &s
}

File: csharp/tests/PublishingMessageTest.cs
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

using System;
using System.Collections.Concurrent;
using System.Text;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using Org.Apache.Rocketmq;

namespace tests
{
    [TestClass]
    public class PublishingMessageTest
    {
        private const string ClientId = "fakeClientId";
        private static readonly Endpoints Endpoints = new Endpoints("127.0.0.1:8081");
        private const string Namespace = "fakeNamespace";


        [TestMethod]
        public void TestNormalMessage()
        {
            const string topic = "yourNormalTopic";
            var message = new Message.Builder().SetTopic(topic).SetBody(Encoding.UTF8.GetBytes("foobar")).Build();
            var topics = new ConcurrentDictionary<string, bool>
            {
                [topic] = true
            };
            var settings = new PublishingSettings(Namespace, ClientId, Endpoints,
                ExponentialBackoffRetryPolicy.ImmediatelyRetryPolicy(3), TimeSpan.FromSeconds(3), topics);
            var publishingMessage = new PublishingMessage(message, settings, false);
            Assert.AreEqual(publishingMessage.MessageType, MessageType.Normal);
        }

        [TestMethod]
        public void TestFifoMessage()
        {
            const string topic = "yourFifoTopic";
            const string messageGroup = "yourMessageGroup";
            var message = new Message.Builder().SetTopic(topic)
                .SetMessageGroup(messageGroup).SetBody(Encoding.UTF8.GetBytes("foobar"))
                .Build();
            var topics = new ConcurrentDictionary<string, bool>
            {
                [topic] = true
            };
            var settings = new PublishingSettings(Namespace, ClientId, Endpoints,
                ExponentialBackoffRetryPolicy.ImmediatelyRetryPolicy(3), TimeSpan.FromSeconds(3), topics);
            var publishingMessage = new PublishingMessage(message, settings, false);
            Assert.AreEqual(publishingMessage.MessageType, MessageType.Fifo);
        }

        [TestMethod]
        public void TestDelayMessage()
        {
            const string topic = "yourDelayTopic";
            var message = new Message.Builder()
                .SetTopic(topic)
                .SetDeliveryTimestamp(DateTime.UtcNow + TimeSpan.FromSeconds(30))
                .SetBody(Encoding.UTF8.GetBytes("foobar")).Build();
            var topics = new ConcurrentDictionary<string, bool>
            {
                [topic] = true
            };
            var settings = new PublishingSettings(Namespace, ClientId, Endpoints,
                ExponentialBackoffRetryPolicy.ImmediatelyRetryPolicy(3),
                TimeSpan.FromSeconds(3), topics);
            var publishingMessage = new PublishingMessage(message, settings, false);
            Assert.AreEqual(publishingMessage.MessageType, MessageType.Delay);
        }

        [TestMethod]
        public void TestTransactionMessage()
        {
            const string topic = "yourTransactionMessage";
            var message = new Message.Builder()
                .SetTopic(topic)
                .SetBody(Encoding.UTF8.GetBytes("foobar")).Build();
            var topics = new ConcurrentDictionary<string, bool>
            {
                [topic] = true
            };
            var settings = new PublishingSettings(Namespace, ClientId, Endpoints,
                ExponentialBackoffRetryPolicy.ImmediatelyRetryPolicy(3),
                TimeSpan.FromSeconds(3), topics);
            var publishingMessage = new PublishingMessage(message, settings, true);
            Assert.AreEqual(publishingMessage.MessageType, MessageType.Transaction);
        }
    }
}